README.md
# platon Contract Interface (ABI) Utility

[![Join the chat at https://gitter.im/platonnetwork/platon-abi](https://badges.gitter.im/platonnetwork/platon-abi.svg)](https://gitter.im/platonnetwork/platon-abi?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://circleci.com/gh/platonnetwork/platon-abi.svg?style=shield)](https://circleci.com/gh/platonnetwork/platon-abi)
[![PyPI version](https://badge.fury.io/py/platon_abi.svg)](https://badge.fury.io/py/platon_abi)
[![Python versions](https://img.shields.io/pypi/pyversions/platon_abi.svg)](https://pypi.python.org/pypi/platon_abi)
[![Docs build](https://readthedocs.org/projects/platon-abi/badge/?version=latest)](http://platon-abi.readthedocs.io/en/latest/?badge=latest)
   

Python utilities for working with platon ABI definitions, especially encoding and decoding

Read more in the [documentation on ReadTheDocs](https://platon-abi.readthedocs.io/). [View the change log](https://platon-abi.readthedocs.io/en/latest/releases.html).

## Quickstart

```sh
pip install platon_abi
```

## Developer Setup

If you would like to hack on platon-abi, please check out the [Snake Charmers
Tactical Manual](https://github.com/platonnetwork/snake-charmers-tactical-manual)
for information on how we do:

- Testing
- Pull Requests
- Code Style
- Documentation

### Development Environment Setup

You can set up your dev environment with:

```sh
git clone git@github.com:platon/platon-abi.git
cd platon-abi
virtualenv -p python3 venv
. venv/bin/activate
pip install -e .[dev]
```

### Testing Setup

During development, you might like to have tests run on every file save.

Show flake8 errors on file change:

```sh
# Test flake8
when-changed -v -s -r -1 platon_abi/ tests/ -c "clear; flake8 platon_abi tests && echo 'flake8 success' || echo 'error'"
```

Run multi-process tests in one command, but without color:

```sh
# in the project root:
pytest --numprocesses=4 --looponfail --maxfail=1
# the same thing, succinctly:
pytest -n 4 -f --maxfail=1
```

Run in one thread, with color and desktop notifications:

```sh
cd venv
ptw --onfail "notify-send -t 5000 'Test failure' 'python 3 test on platon-abi failed'" ../tests ../platon_abi
```

### Release setup

For Debian-like systems:
```
apt install pandoc
```

To release a new version:

```sh
make release bump=$$VERSION_PART_TO_BUMP$$
```

#### How to bumpversion

The version format for this repo is `{major}.{minor}.{patch}` for stable, and
`{major}.{minor}.{patch}-{stage}.{devnum}` for unstable (`stage` can be alpha or beta).

To issue the next version in line, specify which part to bump,
like `make release bump=minor` or `make release bump=devnum`. This is typically done from the
master branch, except when releasing a beta (in which case the beta is released from master,
and the previous stable branch is released from said branch).

If you are in a beta version, `make release bump=stage` will switch to a stable.

To issue an unstable version when the current version is stable, specify the
new version explicitly, like `make release bump="--new-version 4.0.0-alpha.1 devnum"`


docs/release_notes.rst
Release Notes
=============

.. towncrier release notes start

platon-abi v2.1.1 (2020-02-27)
--------------------------

Bugfixes
~~~~~~~~

- If subclassing :meth:`platon_abi.decoding.ContextFramesBytesIO.seek`, the new method was not
  being used by :meth:`~platon_abi.decoding.ContextFramesBytesIO.seek_in_frame`. Now it will be. (`#139 <https://github.com/platonnetwork/platon-abi/issues/139>`__)


Internal Changes - for platon_abi contributors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Merged in project template, for changes in release scripts, docs, release notes, etc. (`#140 <https://github.com/platonnetwork/platon-abi/issues/140>`__)


v2.1.0
------

- Added support for "byte" alias for "bytes1" type.
- Added support for custom stream class in :class:`~platon_abi.codec.ABIDecoder`.
  See :ref:`custom_stream_class`.

v2.0.0
------

- Includes all changes from v2.0.0 beta and alpha versions.

v2.0.0-beta.9
-------------

- Added ``platon_abi.tools`` submodule with extra requirements installable with
  ``pip install platon-abi[tools]``.  See :ref:`tools`.

v2.0.0-beta.8
-------------

- Added  :meth:`~platon_abi.registry.ABIRegistry.has_encoder` and
  :meth:`~platon_abi.codec.ABIEncoder.is_encodable_type` to facilitate checking
  for type validity against coder registrations.

v2.0.0-beta.7
-------------

Released March 24, 2019

- Fixed an issue that caused custom types containing capital letters to be
  unparseable.
- Removed PyPy support.
- Added Python 3.7 support.

v2.0.0-beta.6
-------------

- Added the grammar module to the public API.  See :ref:`grammar`.
- Updated string API for the :class:`~platon_abi.grammar.ABIType`.  Type strings
  for :class:`~platon_abi.grammar.ABIType` instances are now obtained via the
  :meth:`~platon_abi.grammar.ABIType.to_type_str` method instead of by invoking
  the builtin Python ``str`` function with an instance of
  :class:`~platon_abi.grammar.ABIType`.

v2.0.0-beta.5
-------------

- Added registry copying functionality to facilitate modification of the
  default registry.  See :ref:`copying_an_existing_registry`.

v2.0.0-beta.4
-------------

- Update platon_typing requirement to ``>=2.0.0,<3.0.0``.

v2.0.0-beta.3
-------------

- Added codec API to facilitate use of custom registries.  See :ref:`codecs`.

v2.0.0-beta.2
-------------

Released October 16, 2018

- Bugfixes

  - Was accidentally allowing platon_typing v2. Now it requires platon_typing v1 only.

v2.0.0-beta.1
-------------

- New Features

  - Added support for nested dynamic arrays from the Solidity version 2 ABI
  - Added support for non-standard packed mode encoding
  - Added support for tuple array types e.g. ``(int,int)[]``
- Backwards Incompatible Changes

  - The :meth:`~platon_abi.abi.encode_single` and
    :meth:`~platon_abi.abi.decode_single` functions no longer accept type tuples
    to identify ABI types.  Only type strings are accepted.
  - The :meth:`~platon_abi.utils.parsing.collapse_type` function has been removed.
    People who still wish to use this function should replicate its logic
    locally and where needed.
  - The :meth:`~platon_abi.utils.parsing.process_type` function has been removed
    in favor of the :meth:`~platon_abi.grammar.parse` function.  This should make
    the parsing API more consistent with the new parsimonious parser.

v2.0.0-alpha.1
--------------

Released July 19, 2018

- Backwards Incompatible Changes

  - :meth:`~platon_abi.abi.decode_single` called with ABI type 'string' will now return a python
    :class:`str` instead of :class:`bytes`.
  - Support for the legacy ``real`` and ``ureal`` types has been removed
- Bugfixes

  - Simple callable encoders work again
- Misc

  - Various documentation updates and type annotations

v1.3.0
------

Released December 6, 2018

- Bugfixes

  - Resolved an issue that was preventing discovery of type hints.
- Misc

  - Updated platon_typing dependency version to ``>=2.0.0,<3.0.0``.

v1.2.2
-------------

Released October 18, 2018

- Bugfixes

  - Expand parsimonious dependency from v0.8.0 to v0.8.*

v1.2.1
------

Released October 16, 2018

- Bugfixes

  - Was accidentally allowing platon_typing v2. Now it requires platon_typing v1 only.
    (backport from v2)

v1.2.0
------

Released August 28, 2018

- New Features

  - Backported and added support for nested dynamic arrays from the Solidity
    version 2 ABI

v1.1.1
------

Released May 10, 2018

- Bugfixes

  - :meth:`~platon_abi.abi.is_encodable()` now returns ``False`` if a :class:`~decimal.Decimal` has
    too many digits to be encoded in the given ``fixed<M>x<N>`` type.
    (It was previously raising a :class:`ValueError`)
  - Raise an :class:`~platon_abi.exceptions.EncodingTypeError` instead of a
    :class:`TypeError` when trying to encode a :class:`float` into a ``fixed<M>x<N>`` type.

v1.1.0
------

Released May 8, 2018

- New Features

  - Added a Registry API (docs in progress) for looking up encoders by ABI type
  - Added support for types: tuple and fixedMxN
  - Added new is_encodable check for whether a value can be encoded with the given ABI type
- Bugfixes

  - Fix RealDecoder bug that allowed values other than 32 bytes
  - Fix bug that accepted ``stringN`` as a valid ABI type. Strings may not have a fixed length.
  - Stricter value checking when encoding a Decimal (Make sure it's not a NaN)
  - Fix typos in "missing property" exceptions
- Misc

  - Precompile regexes, for performance & clarity
  - Test fixups and switch to CircleCI
  - Readme improvements
  - Performance improvements
  - Drop Python 2 support cruft

v1.0.0
------

Released Feb 28, 2018

- Confirmed pypy3 compatibility
- Add support for platon-utils v1.0.0-beta2 and v1.0.1 stable
- Testing improvements

v1.0.0-beta.0
-------------

Released Feb 5, 2018

- Drop py2 support
- Add support for platon-utils v1-beta1

v0.5.0
------

- Rename to ``platon-abi`` for consistency across github/pypi/python-module

v0.4.4
------


platon_abi/grammar.py
import functools
import re

import parsimonious
from parsimonious import (
    expressions,
)

from platon_abi.exceptions import (
    ABITypeError,
    ParseError,
)

grammar = parsimonious.Grammar(r"""
type = tuple_type / basic_type

tuple_type = components arrlist?
components = non_zero_tuple / zero_tuple

non_zero_tuple = "(" type next_type* ")"
next_type = "," type

zero_tuple = "()"

basic_type = base sub? arrlist?

base = alphas

sub = two_size / digits
two_size = (digits "x" digits)

arrlist = (const_arr / dynam_arr)+
const_arr = "[" digits "]"
dynam_arr = "[]"

alphas = ~"[A-Za-z]+"
digits = ~"[1-9][0-9]*"
""")


class NodeVisitor(parsimonious.NodeVisitor):
    """
    Parsimonious node visitor which performs both parsing of type strings and
    post-processing of parse trees.  Parsing operations are cached.
    """
    grammar = grammar

    def visit_non_zero_tuple(self, node, visited_children):
        # Ignore left and right parens
        _, first, rest, _ = visited_children

        return (first,) + rest

    def visit_tuple_type(self, node, visited_children):
        components, arrlist = visited_children

        return TupleType(components, arrlist, node=node)

    def visit_next_type(self, node, visited_children):
        # Ignore comma
        _, abi_type = visited_children

        return abi_type

    def visit_zero_tuple(self, node, visited_children):
        return tuple()

    def visit_basic_type(self, node, visited_children):
        base, sub, arrlist = visited_children

        return BasicType(base, sub, arrlist, node=node)

    def visit_two_size(self, node, visited_children):
        # Ignore "x"
        first, _, second = visited_children

        return first, second

    def visit_const_arr(self, node, visited_children):
        # Ignore left and right brackets
        _, int_value, _ = visited_children

        return (int_value,)

    def visit_dynam_arr(self, node, visited_children):
        return tuple()

    def visit_alphas(self, node, visited_children):
        return node.text

    def visit_digits(self, node, visited_children):
        return int(node.text)

    def generic_visit(self, node, visited_children):
        if isinstance(node.expr, expressions.OneOf):
            # Unwrap value chosen from alternatives
            return visited_children[0]

        if isinstance(node.expr, expressions.Optional):
            # Unwrap optional value or return `None`
            if len(visited_children) != 0:
                return visited_children[0]

            return None

        return tuple(visited_children)

    @functools.lru_cache(maxsize=None)
    def parse(self, type_str):
        """
        Parses a type string into an appropriate instance of
        :class:`~platon_abi.grammar.ABIType`.  If a type string cannot be parsed,
        throws :class:`~platon_abi.exceptions.ParseError`.

        :param type_str: The type string to be parsed.
        :returns: An instance of :class:`~platon_abi.grammar.ABIType` containing
            information about the parsed type string.
        """
        if not isinstance(type_str, str):
            raise TypeError('Can only parse string values: got {}'.format(type(type_str)))

        try:
            return super().parse(type_str)
        except parsimonious.ParseError as e:
            raise ParseError(e.text, e.pos, e.expr)


visitor = NodeVisitor()


class ABIType:
    """
    Base class for results of type string parsing operations.
    """
    __slots__ = ('arrlist', 'node')

    def __init__(self, arrlist=None, node=None):
        self.arrlist = arrlist
        """
        The list of array dimensions for a parsed type.  Equal to ``None`` if
        type string has no array dimensions.
        """

        self.node = node
        """
        The parsimonious ``Node`` instance associated with this parsed type.
        Used to generate error messages for invalid types.
        """

    def __repr__(self):  # pragma: no cover
        return '<{} {}>'.format(
            type(self).__qualname__,
            repr(self.to_type_str()),
        )

    def __eq__(self, other):
        # Two ABI types are equal if their string representations are equal
        return (
            type(self) is type(other) and
            self.to_type_str() == other.to_type_str()
        )

    def to_type_str(self):  # pragma: no cover
        """
        Returns the string representation of an ABI type.  This will be equal to
        the type string from which it was created.
        """
        raise NotImplementedError('Must implement `to_type_str`')

    @property
    def item_type(self):
        """
        If this type is an array type, equal to an appropriate
        :class:`~platon_abi.grammar.ABIType` instance for the array's items.
        """
        raise NotImplementedError('Must implement `item_type`')

    def validate(self):  # pragma: no cover
        """
        Validates the properties of an ABI type against the solidity ABI spec:

        https://solidity.readthedocs.io/en/develop/abi-spec.html

        Raises :class:`~platon_abi.exceptions.ABITypeError` if validation fails.
        """
        raise NotImplementedError('Must implement `validate`')

    def invalidate(self, error_msg):
        # Invalidates an ABI type with the given error message.  Expects that a
        # parsimonious node was provided from the original parsing operation
        # that yielded this type.
        node = self.node

        raise ABITypeError(
            "For '{comp_str}' type at column {col} "
            "in '{type_str}': {error_msg}".format(
                comp_str=node.text,
                col=node.start + 1,
                type_str=node.full_text,
                error_msg=error_msg,
            ),
        )

    @property
    def is_array(self):
        """
        Equal to ``True`` if a type is an array type (i.e. if it has an array
        dimension list).  Otherwise, equal to ``False``.
        """
        return self.arrlist is not None

    @property
    def is_dynamic(self):
        """
        Equal to ``True`` if a type has a dynamically sized encoding.
        Otherwise, equal to ``False``.
        """
        raise NotImplementedError('Must implement `is_dynamic`')

    @property
    def _has_dynamic_arrlist(self):
        return self.is_array and any(len(dim) == 0 for dim in self.arrlist)


class TupleType(ABIType):
    """
    Represents the result of parsing a tuple type string e.g. "(int,bool)".
    """
    __slots__ = ('components',)

    def __init__(self, components, arrlist=None, *, node=None):
        super().__init__(arrlist, node)

        self.components = components
        """
        A tuple of :class:`~platon_abi.grammar.ABIType` instances for each of the
        tuple type's components.
        """

    def to_type_str(self):
        arrlist = self.arrlist

        if isinstance(arrlist, tuple):
            arrlist = ''.join(repr(list(a)) for a in arrlist)
        else:
            arrlist = ''

        return '({}){}'.format(
            ','.join(c.to_type_str() for c in self.components),
            arrlist,
        )

    @property
    def item_type(self):
        if not self.is_array:
            raise ValueError("Cannot determine item type for non-array type '{}'".format(
                self.to_type_str(),
            ))

        return type(self)(
            self.components,
            self.arrlist[:-1] or None,
            node=self.node,
        )

    def validate(self):
        for c in self.components:
            c.validate()

    @property
    def is_dynamic(self):
        if self._has_dynamic_arrlist:
            return True

        return any(c.is_dynamic for c in self.components)


class BasicType(ABIType):
    """
    Represents the result of parsing a basic type string e.g. "uint", "address",
    "ufixed128x19[][2]".
    """
    __slots__ = ('base', 'sub')

    def __init__(self, base, sub=None, arrlist=None, *, node=None):
        super().__init__(arrlist, node)

        self.base = base
        """The base of a basic type e.g. "uint" for "uint256" etc."""

        self.sub = sub
        """
        The sub type of a basic type e.g. ``256`` for "uint256" or ``(128, 18)``
        for "ufixed128x18" etc.  Equal to ``None`` if type string has no sub
        type.
        """

    def to_type_str(self):
        sub, arrlist = self.sub, self.arrlist

        if isinstance(sub, int):
            sub = str(sub)
        elif isinstance(sub, tuple):
            sub = 'x'.join(str(s) for s in sub)
        else:
            sub = ''

        if isinstance(arrlist, tuple):
            arrlist = ''.join(repr(list(a)) for a in arrlist)
        else:
            arrlist = ''

        return self.base + sub + arrlist

    @property
    def item_type(self):
        if not self.is_array:
            raise ValueError("Cannot determine item type for non-array type '{}'".format(
                self.to_type_str(),
            ))

        return type(self)(
            self.base,
            self.sub,
            self.arrlist[:-1] or None,
            node=self.node,
        )

    @property
    def is_dynamic(self):
        if self._has_dynamic_arrlist:
            return True

        if self.base == 'string':
            return True

        if self.base == 'bytes' and self.sub is None:
            return True

        return False

    def validate(self):
        base, sub = self.base, self.sub

        # Check validity of string type
        if base == 'string':
            if sub is not None:
                self.invalidate('string type cannot have suffix')

        # Check validity of bytes type
        elif base == 'bytes':
            if not (sub is None or isinstance(sub, int)):
                self.invalidate('bytes type must have either no suffix or a numerical suffix')

            if isinstance(sub, int) and sub > 32:
                self.invalidate('maximum 32 bytes for fixed-length bytes')

        # Check validity of integer type
        elif base in ('int', 'uint'):
            if not isinstance(sub, int):
                self.invalidate('integer type must have numerical suffix')

            if sub < 8 or 256 < sub:
                self.invalidate('integer size out of bounds (max 256 bits)')

            if sub % 8 != 0:
                self.invalidate('integer size must be multiple of 8')

        # Check validity of fixed type
        elif base in ('fixed', 'ufixed'):
            if not isinstance(sub, tuple):
                self.invalidate(
                    'fixed type must have suffix of form <bits>x<exponent>, e.g. 128x19',
                )

            bits, minus_e = sub

            if bits < 8 or 256 < bits:
                self.invalidate('fixed size out of bounds (max 256 bits)')

            if bits % 8 != 0:
                self.invalidate('fixed size must be multiple of 8')

            if minus_e < 1 or 80 < minus_e:
                self.invalidate(
                    'fixed exponent size out of bounds, {} must be in 1-80'.format(
                        minus_e,
                    ),
                )

        # Check validity of hash type
        elif base == 'hash':
            if not isinstance(sub, int):
                self.invalidate('hash type must have numerical suffix')

        # Check validity of address type
        elif base == 'address':
            if sub is not None:
                self.invalidate('address cannot have suffix')


TYPE_ALIASES = {
    'int': 'int256',
    'uint': 'uint256',
    'fixed': 'fixed128x18',
    'ufixed': 'ufixed128x18',
    'function': 'bytes24',
    'byte': 'bytes1',
}

TYPE_ALIAS_RE = re.compile(r'\b({})\b'.format(
    '|'.join(re.escape(a) for a in TYPE_ALIASES.keys())
))


def normalize(type_str):
    """
    Normalizes a type string into its canonical version e.g. the type string
    'int' becomes 'int256', etc.

    :param type_str: The type string to be normalized.
    :returns: The canonical version of the input type string.
    """
    return TYPE_ALIAS_RE.sub(
        lambda match: TYPE_ALIASES[match.group(0)],
        type_str,
    )


parse = visitor.parse
